[StructuralMechanicsApplication] Fix hidden adaptation of perturbation size.

dbaumgaertner opened the pull request:
This PR removes a hidden automatism to select the perturbation size in the context of semi-analytic sensitivity analysis and exposes it as an option to the settings. The default behaviour is changed to NOT do automatic adaption unless specified explicitly. The tests are adjusted accordingly. This increases user awareness. 

Beyond its advantages, the hidden adaption becomes problematic, when it is not wanted, but cannot be disabled, such as e.g. with the verification of sensitivities. 

Update: Also the elemental definition of PERTURBATION_SIZE and ADAPT_PERTURBATION_SIZE was removed.

dbaumgaertner committed:
added new variable

dbaumgaertner committed:
removed hidden automatism

dbaumgaertner committed:
added new option to relevant test

behavior and results of tests are not altered

philbucher commented on applications/StructuralMechanicsApplication/tests/response_function_tests/adjoint_stress_response_parameters.json:
```diff
@@ -13,6 +13,7 @@
         "response_type"    : "adjoint_local_stress",
         "gradient_mode"     : "semi_analytic",
         "step_size"         : 1.0e-5,
+        "adapt_step_size"  : true,
```
please make the naming consistent, why is the setting called `adapt_step_size` and the variable `APPLY_ADAPTIVE_PERTURBATION`?

armingeiser replied:
The step_size/perturbation size difference is already there.
I would vote for `adapt_step_size"` in the json settings and `ADAPT_PERTURBATION_SIZE` for the variable name.
If the json settings are changed from `"step_size"` to `"perturbation_size"`, it will change many existing examples...

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp:
```diff
@@ -544,16 +544,30 @@ void AdjointFiniteDifferencingBaseElement::CalculateStressDesignVariableDerivati
+        delta = this->GetValue(PERTURBATION_SIZE) * correction_factor;
+    }
+    else
+        delta = this->GetValue(PERTURBATION_SIZE);
```
I think this can be simplified, please come to speak in person, thats easier :)

armingeiser replied:
:+1:

philbucher reviewed (commented):
I have some suggestions, can we please speak in person?

MFusseder commented:
I am fine with the introduction of this perturbation mechanism.

armingeiser reviewed (changes requested):
Thanks for doing this, i have more or less the same comments as @philbucher

dbaumgaertner committed:
Merge branch 'master' into fix-adaptive-perturbation

dbaumgaertner committed:
Merge branch 'master' into fix-adaptive-perturbation

dbaumgaertner committed:
adjusted variable name

dbaumgaertner committed:
trigger adaption of step size through process info

dbaumgaertner committed:
adjusted tests to include step adaption explicitly

dbaumgaertner committed:
Set and get perturbation size through process info

Hence no need anymore for an elemental definition of this variable

dbaumgaertner committed:
adjusted tests to use perturbation size from process info

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.h:
```diff
@@ -458,12 +458,12 @@ class AdjointFiniteDifferencingBaseElement : public Element
      * Get the perturbation size for a scalar variable
      */
-    double GetPerturbationSize(const Variable<double>& rDesignVariable);
+    double GetPerturbationSize(const Variable<double>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo);
```
```suggestion
    double GetPerturbationSize(const Variable<double>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo) const;
```

dbaumgaertner commented:
@armingeiser @philbucher @MFusseder I addressed all your comments. Additionally, I implemented the hint of @philbucher to store some variables in the process info rather on each element/condition separately. 

All tests are running. Now it's ready to be reviewed again.

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp:
```diff
@@ -555,19 +555,33 @@ void AdjointFiniteDifferencingBaseElement<TPrimalElement>::CalculateStressDesign
+    }
+    else
+        delta = rCurrentProcessInfo[PERTURBATION_SIZE];
+
```
recommendation:
~~~c
double delta = rCurrentProcessInfo[PERTURBATION_SIZE];
if (rCurrentProcessInfo[ADAPT_PERTURBATION_SIZE]) {
        delta *= this->GetPerturbationSizeModificationFactor(rDesignVariable);
}
~~~

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp:
```diff
@@ -555,19 +555,33 @@ void AdjointFiniteDifferencingBaseElement<TPrimalElement>::CalculateStressDesign
+    }
+    else
+        delta = rCurrentProcessInfo[PERTURBATION_SIZE];
+
```
same

philbucher reviewed (commented):
much better!

some minor comments

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.h:
```diff
@@ -458,12 +458,12 @@ class AdjointFiniteDifferencingBaseElement : public Element
      * Get the perturbation size for a vector variable
      */
-    double GetPerturbationSize(const Variable<array_1d<double,3>>& rDesignVariable);
+    double GetPerturbationSize(const Variable<array_1d<double,3>>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo);
```
```suggestion
    double GetPerturbationSize(const Variable<array_1d<double,3>>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo) const;
```

dbaumgaertner replied:
Done. Thanks for the hint.

philbucher commented on applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_structural_response_function.cpp:
```diff
@@ -52,8 +49,13 @@ namespace Kratos
-            VariableUtils().SetNonHistoricalVariable(PERTURBATION_SIZE, mDelta, mrModelPart.Elements());
-            VariableUtils().SetNonHistoricalVariable(PERTURBATION_SIZE, mDelta, mrModelPart.Conditions());
+            double delta = mResponseSettings["step_size"].GetDouble();
+            mrModelPart.GetProcessInfo()[PERTURBATION_SIZE] = delta;
```
here the naming is inconsistent
delta & PERTURBATION_SIZE & step_size

dbaumgaertner replied:
I changed the delta, which indeed does not make sense. About the difference PERTURBATION_SIZE and step_size, see @armingeiser comment above.

dbaumgaertner committed:
minor naming

dbaumgaertner committed:
added const functions to avoid conflicts

dbaumgaertner committed:
minor refactor after review

armingeiser reviewed (approved):
Thanks for this!
